Expand integer range

haguenau opened the pull request:
The published version of ngrep appears to have to small issues:

* The range for many integral options is small because the underlying type is a 16-bit unsigned (so it's impossible to ask to capture 100000 packets using `-n`, for example);

* the option-parsing code does not check for overflow conditions, resulting in silent wraparound (so that asking to capture exactly 128K packets will be the same as `-n 0`: no limit to how many packets will be captured).

This change set:

* switches the type for many counters to unsigned long, so that it's at least 32 bits;
* checks command line values for syntactic correctness and range, and errors out if needed.

haguenau commented:
By the way, I'm completely open to altering the change set to e.g. switch the integral type to a known-width one instead of unsigned long, if that makes it a better fit to the existing coding style.

jpiccari commented on ngrep.c:
```diff
@@ -105,9 +106,10 @@
+
 #if USE_TCPKILL
-uint16_t tcpkill_active = 0;
+unsigned long tcpkill_active = 0;
```
This is a bad idea. `tcpkill_active` is used as the third parameter to a routine from ported from dSniff called [`tcpkill_kill`](https://github.com/haguenau/ngrep/commit/f2a1f8b09194a8ab87787c1fdb60b968f5957c05#diff-a3cce3c8aad81bc94746a67bc7a0b5e2R54). It expects the third parameter to be `unsigned`, but ultimately gets compared against an `int`. So the real max value that can be passed with -K is `sizeof(int)` (aka `uint16_t`).

Probably best to leave this one as is.

jpiccari commented on ngrep.c:
```diff
@@ -105,9 +106,10 @@
 
 uint16_t snaplen = 65535, limitlen = 65535, promisc = 1, to = 100;
-uint16_t match_after = 0, keep_matching = 0, matches = 0, max_matches = 0;
+unsigned long match_after = 0, keep_matching = 0, matches = 0, max_matches = 0;
```
IIRC `unsigned long` on most PC hardware will be 32-bit but that is not guaranteed. Instead you should use the C99 fixed width types. In this case either `uint32_t` or `uint64_t`. This will also keep things consistent with the rest of the file.

jpiccari commented on ngrep.c:
```diff
@@ -1557,4 +1573,27 @@ char *win32_choosedevice(void) {
+}
+
+void dierange(char *msg, unsigned long value) {
+    fprintf(stderr, "`%s': value `%lu' is out of range.\n", msg, value);
```
If this routine expected a character array as the second parameter I think the feedback would be more valuable since in some cases we won't be able to output `value` properly as an `unsigned long`.

haguenau replied:
* Joshua Piccari <notifications@github.com>, 2015-08-21 23:10:56 Fri:
> > +void dierange(char *msg, unsigned long value) {
> > +    fprintf(stderr, "`%s': value `%lu' is out of range.\n", msg, value);
> If this routine expected a character array as the second parameter I
> think the feedback would be more valuable since in some cases we
> won't be able to output `value` properly as an `unsigned long`.

I believe the combination of `parseulong` (which errors out if the input is syntactically invalid) and `dierange` (which is called specifically for syntactically correct, but out of range integers) covers all cases that are relevant for this change set. I think that's what you are referring to in your edit.

haguenau commented:
Thanks for the review! I'll try to find time to respond to the comments by updating the code soon.

haguenau commented:
@jpr5, @jpiccari: I switched the underlying type for most parameters to `uint64_t`, meaning counters will go 64-bit on all platforms. I also switched the tcpkill variable to `unsigned`, since this is the type that `tcpkill_kill` expects.

jpiccari commented:
:+1: I'll get this merged in. Thanks for making the updates!

jpiccari commented:
Sorry it took so long, I cleaned up the commit history to be more concise and adjusted the code style to fit in better with the existing stuff. Also, I updated tcpkill_kill() to expect 32-bit values for the kill_count.

Thanks for contributing!
